### PR TITLE
feat: enhance visualization of connectors

### PR DESCRIPTION
**Pull request by haoqixu**
partially implement #117

This PR enhances visualization of connectors:
- The connection between connector's exporter and receiver is visualised as a directed arrows from exporter to receiver.
- Pipeline Nodes (parentNode) are laid out based on the connector's connections. This is different from what was mentioned in the issue.

before:
图片

after:
图片

### CODE REVIEW

**Comment by haoqixu** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by mmanciop** [comment]
Lovely :-) We'll review it ASAP!

**Comment by mmanciop** [comment]
@haoqixu would you please enable in your fork [Allowing changes to a pull request branch created from a fork](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork), or run `(cd packages/otelbin/ && npm run prettier:all)` from the repo root? The linter is unhappy because of missing semicolons :-)

**Comment by mmanciop** [comment]
Overall, I like very much where the change is going. The layout you implemented has a lot going for it: it is easier to see how pipelines flow into one another, following the conventional left-to-right approach.

I have concerns about the fact that it is no longer simple to see which pipelines can receive data "from the outside" through non-connector receivers, which exports send data to the outside, which is why in  #117 I was thinking of always keeping lanes vertically aligned.

Nevertheless, this looks like a marked improvement to me when dealing with connectors, and I am looking forward for @bripkens to review when he is no longer under the weather.

**Comment by mmanciop** [comment]
I am also not sure how far it is realistic to go in terms of detecting connector cycles ([see example](https://www.otelbin.io/s/e8f437fb-67b8-4622-8620-f19e3f9f005b)), but we definitely should try not to have edges overlap with nodes.
Screenshot 2023-12-12 at 09 28 31

**Comment by haoqixu** [comment]
> I am also not sure how far it is realistic to go in terms of detecting connector cycles ([see example](https://www.otelbin.io/s/e8f437fb-67b8-4622-8620-f19e3f9f005b)), but we definitely should try not to have edges overlap with nodes.

I think it is realistic to detect cycles and I have implemented a POC. It looks like:

图片
图片

After I clean up the code and test it more carefully, I will update the PR.

**Comment by haoqixu** [comment]
I have rebase this PR onto the latest main and add the code for cycles detection. Here are some snapshots after update:
图片
图片

File: packages/otelbin/src/components/react-flow/CyclicErrorEdge.tsx [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+			<EdgeLabelRenderer>
+				<span
+					style={{
+						position: "absolute",
+						transform: `translate(50%, -100%) translate(${labelX}px, ${labelY}px)`,
+						pointerEvents: "all",
+						...style,
+						color: hovered ? "var(--color-critical-hover)" : style?.color,
+					}}
+					onMouseEnter={() => setHovered(true)}
```
Comment (bripkens) [inline packages/otelbin/src/components/react-flow/CyclicErrorEdge.tsx:60]: **Suggestion:** Let's remove this hover effect. We also don't have this for the other edge types.

File: packages/otelbin/src/components/react-flow/CyclicErrorEdge.tsx [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+					onMouseLeave={() => setHovered(false)}
+				>
+					CyclicError
```
Comment (bripkens) [inline packages/otelbin/src/components/react-flow/CyclicErrorEdge.tsx:63]:
```suggestion
					Cycle in pipeline
```

File: packages/otelbin/src/components/react-flow/layout/getSimpleCycles.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+// The types of dagre API is incorret so we have to disable type check for this file.
+//
+// eslint-disable-next-line
+// @ts-nocheck
+
+import Dagre from "@dagrejs/dagre";
+
+// getSimpleCycles use Johnson algorithm to find all the simple cycles of the graph.
+// ref: https://www.cs.tufts.edu/comp/150GA/homeworks/hw1/Johnson%2075.PDF
+export function getSimpleCycles(g: Dagre.graphlib.Graph) {
```
Comment (bripkens) [inline packages/otelbin/src/components/react-flow/layout/getSimpleCycles.ts:13]: **Question:** For licensing reasons, is this a code a copy from some location? It somehow looks like it, especially with the different casing of `johnson_simple_cycles`?

**Review by bripkens — COMMENTED** [review]
This is looking really good @haoqixu. Great job! ❤️ 

Left some minor comments about the code.

Could you please visually align the arrow head of the cycle indicator edge to the arrow heads of the other edge types? That would be great.

image

File: packages/otelbin/src/components/react-flow/layout/getSimpleCycles.ts (hunk 3 above)
Comment (haoqixu) [reply packages/otelbin/src/components/react-flow/layout/getSimpleCycles.ts:13]: It is not copied from somewhere else. I found some python and C++ reference code and this code is basically written based on the pseudo code in reference pdf.

**Comment by haoqixu** [comment]
updated:
图片

File: packages/otelbin/src/components/react-flow/CyclicErrorEdge.tsx [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+			<EdgeLabelRenderer>
+				<span
+					style={{
+						position: "absolute",
+						transform: `translate(50%, -100%) translate(${labelX}px, ${labelY}px)`,
+						pointerEvents: "all",
+						...style,
+					}}
+				>
+					Cycle in pipeline
```
Comment (mmanciop) [inline packages/otelbin/src/components/react-flow/CyclicErrorEdge.tsx:57]:
```suggestion
					connector cycle
```

**Review by bripkens — APPROVED** [review]
LGTM 👍 

Really well done! Thank you @haoqixu!

**Pull request merged**